collection.go

```diff
@@ -426,9 +445,19 @@ func (cl *collectionLoader) loadMap(mapName string) (*Map, error) {
 		return nil, fmt.Errorf("map %s: BTF doesn't match collection", mapName)
 	}
 
-	m, err := newMapWithOptions(mapSpec, cl.opts.Maps, cl.handles)
-	if err != nil {
-		return nil, fmt.Errorf("map %s: %w", mapName, err)
+	var m *Map
+	var err error
+	var ok bool
+
+	if m, ok = cl.opts.OverrideMaps[mapName]; ok {
+		if err := mapSpec.checkCompatibility(m); err != nil {
```

Comment:
You can move this check into newCollectionLoader and then initialise cl.maps with opts.OverrideMaps.

Reply:
What should happen to the maps in opts.OverrideMaps when a call to `NewCollectionWithOptions()` / `LoadAndAssign()` fails? Should they be closed? I'm asking because if we initialize `cl.maps` with `opts.OverrideMaps` those will be closed in case of error in `(cl *collectionLoader) cleanup()`.

Reply:
> What should happen to the maps in `opts.OverrideMaps` when a call to `NewCollectionWithOptions()` / `LoadAndAssign()` fails? Should they be closed?

Probably not. We don't own those Maps, and it will cause bugs in user code (and cause data loss if not pinned).

> those will be closed in case of error in `(cl *collectionLoader) cleanup()`.

Still the case in the current proposal, right? You're pulling a Map from OverrideMaps and putting it into `cl.maps` below.

I agree with having the compatibility check here, but just make `loadMap()` early-return the Map here if it exists in the overrides, without assigning it to `cl.maps`. The overrides then effectively serve as a second Map cache.

Reply:
Good catch re: what happens on error. This leads to the follow up question: what should happen if I use `NewCollectionWithOptions` directly (so not via `LoadAndAssign`)? In that case `coll.Maps` will contain references to `Replacements`. Who owns those? Maybe we need to `Clone()` the replacement and stick it in cl.maps after all?

Reply:
> Still the case in the current proposal, right? You're pulling a Map from OverrideMaps and putting it into cl.maps below.

Yes, almost the same. The small difference is that I'm not pulling all maps at once. 

> Who owns those?

It's a good question, not only in the error case but in general: 

```
m := ebpf.NewMap(...)
col := ebpf.NewCollectionWithOptions(..., ebpf.CollectionOptions{MapReplacements: ... "mymap": m})

....

col.Close()

// should m be valid here? 
```

> Maybe we need to Clone() the replacement and stick it in cl.maps after all?

It seems it'd simplify things.

Reply:
I've trying to reproduce this problem but I can't.  

> since loadMaps isn't called for all maps when using LoadAndAssign.

If loadMaps is not called it means the map is not used, hence not replace and there's nothing to worry about its compatibility. What I am missing?

Reply:
> If loadMaps is not called it means the map is not used

Correct.

> What I am missing?

Lorenz wants to make sure all given MapReplacements are always compatibility-checked, regardless of whether they are loaded or not. This avoids any surprises.

I'll pick this up in a follow-up together with the ReplaceMaps() deprecation.